### PR TITLE
Skip internal distributed tables in schema_change_test

### PR DESCRIPTION
**Pull request by psarna**
The original idea for `schema_change_test` was to ensure that **if** schema hasn't changed, the digest also remained unchanged. However, a cumbersome side effect of adding an internal distributed table (or altering one) is that all digests in `schema_change_test` are immediately invalid, because the schema changed.
Until now, each time a distributed system table was added/amended, a new test case for `schema_change_test` was generated, but this effort is not worth the effect - when a distributed system table is added, it will always propagate on its own, so generating a new test case does not bring any tangible new test coverage - it's just a pain.

To avoid this pain, `schema_change_test` now explicitly skips all internal keyspaces - which includes internal distributed tables - when calculating schema digest. That way, patches which change the way of computing the digest itself will still require adding a new test case, which is good, but, at the same time, changes to distributed tables will not force the developers to introduce needless schema features just for the sake of this test.

Tests:
 * unit(dev)
 * manual(rebasing on top of a change which adds two distributed system tables - all tests still passed)

Refs #7617

### CODE REVIEW

File: db/schema_tables.cc [hunk 1]
```diff
@@ -695,16 +695,16 @@ redact_columns_for_missing_features(mutation m, schema_features features) {
  * Read schema from system keyspace and calculate MD5 digest of every row, resulting digest
  * will be converted into UUID which would act as content-based version of the schema.
  */
-future<utils::UUID> calculate_schema_digest(distributed<service::storage_proxy>& proxy, schema_features features)
+future<utils::UUID> calculate_schema_digest(distributed<service::storage_proxy>& proxy, schema_features features, noncopyable_function<bool(std::string_view)> skip_predicate)
```
Comment (avikivity) [inline db/schema_tables.cc:698]: Maybe the positive form would be nicer for understanding, include_predicate.

File: test/boost/schema_change_test.cc [hunk 2]
```diff
@@ -637,7 +637,13 @@ future<> test_schema_digest_does_not_change_with_disabled_features(sstring data_
...
 
         auto expect_digest = [&] (schema_features sf, utils::UUID expected) {
-            auto actual = calculate_schema_digest(service::get_storage_proxy(), sf).get0();
+            // Changes in the system_distributed keyspace was a common source of regenerating this test case.
+            // It was not the original idea of this test to regenerate after each change in the distributed
+            // system keyspace, since these changes are propagated naturally on their own anyway - the schema
+            // with highest timestamp will win and be sent to other nodes.
+            // Thus, system_distributed.* tables are officially not taken into account,
+            // which makes it less likely that this test case would need to be needlessly regenerated.
+            auto actual = calculate_schema_digest(service::get_storage_proxy(), sf, &is_internal_keyspace).get0();
```
Comment (avikivity) [inline test/boost/schema_change_test.cc:646]: Maybe a more positive way is to specify the keyspace for our test tables (`tests`). This way we don't depend on the muddy definition of is_internal_keyspace.

On the other hand, a positive way means that if we misspell the keyspace name, the test does nothing.

Comment (avikivity) [reply test/boost/schema_change_test.cc:646]: (A positive way could even be exposed: calculate a digest for this keyspace with a REST API or cql function; this way we could run the tests manually for a selected schema:

```sql
cql> SELECT keyspace_schema_digest('my_keyspace')
blah-blah-blah
```

So you can pre-test interactively before going to schema_change_test to formalize the test.

**Comment by avikivity** [comment]
Looks good, minor comments.

File: test/boost/schema_change_test.cc (hunk 2 above)
Comment (psarna) [reply test/boost/schema_change_test.cc:646]: On the other hand, if somebody adds an internal keyspace and forgets to mark it as internal, this test will incidentally catch that and report a failed schema. So perhaps it still makes sense to digest all non-internal tables here.

File: db/schema_tables.cc (hunk 1 above)
Comment (psarna) [reply db/schema_tables.cc:698]: Sure, both are equally fine with me, which makes the positive form win.

Comment (psarna) [reply db/schema_tables.cc:698]: One slight disadvantage of the positive form is that I can no longer pass a plain function pointer as a predicate, and instead have to negate it just to negate it back inside the `calculate_schema_digest` implementation. It's not a big issue though.

**Comment by psarna** [comment]
v2:
 - changed the predicate to be positive - it now calculates the digest only for keyspaces which match the predicate

For now I'm leaving the use of `is_internal_keyspace` in the test as is, while waiting for feedback of https://github.com/scylladb/scylla/pull/7854#discussion_r551166273

File: db/schema_tables.cc [hunk 3]
```diff
@@ -695,16 +695,16 @@ redact_columns_for_missing_features(mutation m, schema_features features) {
  * Read schema from system keyspace and calculate MD5 digest of every row, resulting digest
  * will be converted into UUID which would act as content-based version of the schema.
  */
-future<utils::UUID> calculate_schema_digest(distributed<service::storage_proxy>& proxy, schema_features features)
+future<utils::UUID> calculate_schema_digest(distributed<service::storage_proxy>& proxy, schema_features features, noncopyable_function<bool(std::string_view)> accept_predicate)
```
Comment (avikivity) [inline db/schema_tables.cc:698]: I'll merge, but the function signature for the predicate should have mentioned that it's the keyspace name being tested (either in the name of the predicate or the parameter).

File: db/schema_tables.hh [hunk 4]
```diff
@@ -166,6 +166,7 @@ future<> save_system_schema(cql3::query_processor& qp, const sstring & ks);
 // saves/creates "system_schema" keyspace
 future<> save_system_keyspace_schema(cql3::query_processor& qp);
 
+future<utils::UUID> calculate_schema_digest(distributed<service::storage_proxy>& proxy, schema_features, noncopyable_function<bool(std::string_view)> accept_predicate);
 future<utils::UUID> calculate_schema_digest(distributed<service::storage_proxy>& proxy, schema_features);
```
Comment (kbr-) [inline db/schema_tables.hh:170]: Maybe a comment explaining that this calculates digest using non-system keyspaces?

File: db/schema_tables.hh [hunk 5]
```diff
@@ -166,6 +166,7 @@ future<> save_system_schema(cql3::query_processor& qp, const sstring & ks);
 // saves/creates "system_schema" keyspace
 future<> save_system_keyspace_schema(cql3::query_processor& qp);
 
+future<utils::UUID> calculate_schema_digest(distributed<service::storage_proxy>& proxy, schema_features, noncopyable_function<bool(std::string_view)> accept_predicate);
```
Comment (kbr-) [inline db/schema_tables.hh:169]: Maybe a comment explaining that the predicate tests keyspaces?

**Review by kbr- — APPROVED** [review]
Two nits, otherwise LGTM

**Pull request closed without merging**